File: Algorithms/Python3.x/859-Buddy_Strings.py
# Runtime: 44 ms, faster than 31.29% of Python3 online submissions for Buddy Strings.
# Memory Usage: 14.1 MB, less than 8.00% of Python3 online submissions for Buddy Strings.
class Solution:
    def buddyStrings(self, A: str, B: str) -> bool:
        if len(A) != len(B):
            return False
        if A == B and len(set(A)) < len(B):
            return True
        diff = [(a, b) for a, b in zip(A, B) if a != b]
        return len(diff) == 2 and diff[0] == diff[1][::-1]

File: Algorithms/C++/859-Buddy_Strings.cpp
// Runtime: 12 ms, faster than 9.24% of C++ online submissions for Buddy Strings.
// Memory Usage: 9.1 MB, less than 87.50% of C++ online submissions for Buddy Strings.
class Solution {
public:
    bool buddyStrings(string A, string B) {
        if (A.size() != B.size())
            return false;
        if (A == B) {
            set<char> seen;
            for (char a : A) {
                if (seen.find(a) != seen.end())
                    return true;
                seen.insert(a);
            }
            return false;
        } else {
            vector<pair<char, char>> diff;
            for (int i = 0; i < A.size(); ++i) {
                if (A[i] != B[i])
                    diff.push_back(make_pair(A[i], B[i]));
                if (diff.size() > 2)
                    return false;
            }
            return diff.size() == 2 && diff[0].first == diff[1].second && diff[0].second == diff[1].first;
        }
    }
};
